src/fs/fs.c

```diff
@@ -387,6 +388,16 @@ int cfs_fs_create(const str256_t *fs_name)
 		      &fs_node->cfs_fs);
 
 	LIST_INSERT_HEAD(&fs_list, fs_node, link);
+
+	/* 
+	 * We have made a copy of namespace and kvtree.
+	 * free the data structures else they will leak.
+	 */
+	kvstor = kvstore_get();
```

Comment:
There are certain failure condition like failure of `fs_node_init` will jump to `delete_kvtree` lable where `kvtree` and `ns` won't be freed. So just to avoid that you can initialize both `kvtree = NULL` and `ns = NULL` and on last label you put if condition for free which will make sure we always free them in any condition.

if ( kvtree != NULL ) {
    kvs_free(kvstore, kvtree);
}

Reply:
Jumping to delete_kvtree label will call kvtree_delete() and ns_delete(), these functions will free kvtree and ns respectively.
I am not doing any code change now. Please check, if you have doubt, we can read the code together.

Reply:
Got it thank You.